### PR TITLE
Add `HashEngine` trait for writing bytes without dealing with impossible io::Error returns

### PR DESCRIPTION
**Pull request by apoelstra**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -44,6 +44,17 @@ use std::{fmt, io, ops};
 
 pub use error::Error;
 
+/// A hashing engine which bytes can be serialized into. It is expected
+/// to implment the `io::Write` trait, but to never return errors under
+/// any conditions.
+pub trait HashEngine: Clone + io::Write {
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply src/lib.rs]: So that someone can reuse a prefix when hashing.

File: src/lib.rs [hunk 2]
```diff
@@ -44,6 +44,17 @@ use std::{fmt, io, ops};
...
+    /// Add data to the hash engine without any error return type to deal with
+    #[inline(always)]
+    fn input(&mut self, data: &[u8]) {
+        self.write(data).expect("hash returned error");
```
Comment (dongcarl) [inline src/lib.rs]: Since `write()` does not guarantee that all bytes are written, we should use `write_all()`.

**Comment by dongcarl** [comment]
@apoelstra I think Travis is complaining about `HashEngine` not being implemented for SHA512, needs rebase.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (apoelstra) [reply src/lib.rs]: I don't think it's reasonable to make this less ergonomic to support people who don't read documentation.

**Comment by apoelstra** [comment]
Rebased.

**Comment by apoelstra** [comment]
Added function to expose the midstate. @sgeisler do you think this requires enough thinking now that people won't do reckless things?

**Comment by apoelstra** [comment]
Yep, you're welcome to open such issues. Though associated constants are such a big deal that I doubt we'd forget them for this library :).

**Pull request merged**